quic/s2n-quic-core/src/recovery/bbr/probe_rtt.rs

```diff
@@ -0,0 +1,147 @@
+    timer: Timer,
+    round_done: bool,
+}
+
+/// Methods related to the ProbeRtt state
+impl BbrCongestionController {
+    /// Check if it is time to start probing for RTT changes, and enter the ProbeRtt state if so
+    pub fn check_probe_rtt<Rnd: random::Generator>(
+        &mut self,
+        bytes_in_flight: u32,
+        random_generator: &mut Rnd,
+        now: Timestamp,
+    ) {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.3.4.4
+        //# BBRCheckProbeRTT()
+
+        // TODO: check BBR.state != ProbeRtt
+        if self.data_volume_model.probe_rtt_expired() && !self.idle_restart {
+            // TODO: self.state = ProbeRtt
+            self.save_cwnd();
+            self.probe_rtt_state.timer.cancel();
+            self.probe_bw_state.ack_phase = AckPhase::ProbeStopping;
+            self.round_counter
+                .set_round_end(self.bw_estimator.delivered_bytes());
+        }
+
+        // TODO: if BBR.state == ProbeRtt
+        self.handle_probe_rtt(bytes_in_flight, random_generator, now);
+
+        if self.bw_estimator.rate_sample().delivered_bytes > 0 {
+            self.idle_restart = false;
+        }
+    }
+
+    /// Keeps BBR in the ProbeRTT state for max of (PROBE_RTT_DURATION, 1 round)
+    ///
+    /// ProbeRTT will be exited when this threshold is reached
+    fn handle_probe_rtt<Rnd: random::Generator>(
+        &mut self,
+        bytes_in_flight: u32,
+        random_generator: &mut Rnd,
+        now: Timestamp,
+    ) {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.3.4.4
+        //# BBRHandleProbeRTT()
+
+        // Ignore low rate samples during ProbeRTT
+        self.bw_estimator.on_app_limited(bytes_in_flight);
+
+        if !self.probe_rtt_state.timer.is_armed() && bytes_in_flight <= self.probe_rtt_cwnd() {
+            // Wait for at least ProbeRTTDuration to elapse:
+            self.probe_rtt_state.timer.set(now + PROBE_RTT_DURATION);
+            // Wait for at least one round to elapse:
+            self.probe_rtt_state.round_done = false;
+            self.round_counter
+                .set_round_end(self.bw_estimator.delivered_bytes());
+        } else if self.probe_rtt_state.timer.is_armed() {
+            if self.round_counter.round_start() {
+                self.probe_rtt_state.round_done = true;
+            }
+            if self.probe_rtt_state.round_done {
+                self.check_probe_rtt_done(random_generator, now)
+            }
+        }
+    }
+
+    /// Checks if the ProbeRtt state should be exited and calls `exit_probe_rtt` if so
+    ///
+    /// The next ProbeRtt will be scheduled as necessary
+    pub fn check_probe_rtt_done<Rnd: random::Generator>(
+        &mut self,
+        random_generator: &mut Rnd,
+        now: Timestamp,
+    ) {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.3.4.4
+        //# BBRCheckProbeRTTDone()
```

Comment:
I think including all of the pseudocode might help in checking functional equivalence. The citation as it is, just acknowledges the fact that the function `BBRCheckProbeRTTDone` exists, which isn't the interesting part IMO

Reply:
Yeah, I've gone back and forth on that. I've been including just the name, as I didn't want to be _too_ tied to how the pseudocode implements things, but I agree that when its pretty close it does help to verify the equivalence. I can try it out